### PR TITLE
Basing Body.string on stream instead of reader

### PR DESCRIPTION
**Pull request by kareblak**
To enable specifying charset in the Body utility. Wrapping stream in a loaner to swallow the possible exception. Rewrote test to run utf-8 since we're all better off making this default.

### CODE REVIEW

**Comment by softprops** [comment]
+1 for me. @n8han ?

**Comment by kareblak** [comment]
It might come as a surprise for people already relying on the possible iso-8859-1 decoding, but that shouldn't really go as an assumption anymore imo.

File: library/src/main/scala/request/readers.scala [hunk 1]
```diff
@@ -17,17 +21,22 @@ object Body {
...
+      }
+      read()
+      writer.toString
     }
-    read()
-    reader.close
-    writer.toString
+  }
+  
+  def loan[C <: Closeable, B](c: => C)(f: (C) => B): B = {
```
Comment (n8han) [inline library/src/main/scala/request/readers.scala]: There is a thing similar to this, `unfiltered.util.IO.use`

**Comment by softprops** [comment]
> It might come as a surprise for people already relying on the possible iso-8859-1 decoding

I think its a more common case that it has come as an every bigger surprise when you _expect_ UTF8 :)

**Comment by kareblak** [comment]
Very much so indeed!

**Comment by kareblak** [comment]
@n8han changed it to IO.use

**Comment by n8han** [comment]
Would you say that an underlying problem here is that `HttpRequest#reader` does not allow you to specify a character set?

**Comment by kareblak** [comment]
Yes. But not so much as iso-8859-1 being the default charset, imo. At least on my utf-8 systems. And in the tests.

**Comment by kareblak** [comment]
Wanna solve this with ```HttpRequest#reader ``` or do it this way temporarily?

**Comment by n8han** [comment]
> Wanna solve this with HttpRequest#reader or do it this way temporarily?

Yeah, sorry for the slow back and forth. I'm hesitant to merge as we have introduced regressions in the past through changes to the input interfaces. They can be overridden and are filtered for gzip compression, so it's easy for a change to have unintended consequences for those use cases. I'm a little gun shy because of that.

This is also a behavioral change and if someone is relying on the existing default, to use the servlet container's default character set for reading the request, they'll have a breakage until they add in a parameter request what was previously their default.

I'd be more comfortable with a change that makes it possible to specify the character set here without changing the default, which I believe would be the JVM default (unfortunately, I think the java stdlib methods that use the default are all deprecated?).

**Comment by omarkilani** [comment]
@kareblak are you still interested in this PR?

**Comment by omarkilani** [comment]
Please re-open this if required.

**Pull request closed without merging**